### PR TITLE
Fix steering behaviors implementation and start using physics environments

### PR DESCRIPTION
**Pull request by mazzio97**
- Influence spheres now use the shapes of alchemist-implementationbase
- Introduction of steering strategies in order to create different methods for combining the steering actions together
- Environments hierarchy changed: now Continuous2DEnvironment extends EuclideanPhysics2DEnvironmentImpl

### CODE REVIEW

**Comment by DanySK** [comment]
First and foremost, it must pass on CI

File: alchemist/alchemist-cognitive-agents/src/main/kotlin/it/unibo/alchemist/model/cognitiveagents/characteristics/cognitive/AbstractCognitiveCharacteristic.kt [hunk 1]
```diff
@@ -6,6 +6,13 @@ import it.unibo.alchemist.model.cognitiveagents.characteristics.PARAMETERS_FILE
 abstract class AbstractCognitiveCharacteristic : CognitiveCharacteristic {
 
     protected var currentLevel: Double = 0.0
+        set(value) {
+            field = when {
```
Comment (DanySK) [inline alchemist/alchemist-cognitive-agents/src/main/kotlin/it/unibo/alchemist/model/cognitiveagents/characteristics/cognitive/AbstractCognitiveCharacteristic.kt]: isn't this a very counterintuitive way for `min(1.0, max(0.0, value))`? by the way, there is probably a library function to constrain values

File: alchemist/alchemist-cognitive-agents/src/main/kotlin/it/unibo/alchemist/model/cognitiveagents/characteristics/cognitive/Fear.kt [hunk 2]
```diff
@@ -20,5 +20,5 @@ class Fear(
     )
 
     private fun List<CognitivePedestrian<*>>.aggregateFears() =
-        this.sumByDouble { it.fear() } / this.size
+        if (size > 0) this.sumByDouble { it.fear() } / this.size else 0.0
```
Comment (DanySK) [inline alchemist/alchemist-cognitive-agents/src/main/kotlin/it/unibo/alchemist/model/cognitiveagents/characteristics/cognitive/Fear.kt]: If you don't mind use curly brackets. It's hard to parse visually

**Comment by mazzio97** [comment]
The only error I can notice in the CI is this one but it's strange because I've never touched neither AwtEuclidean2DShapeFactory nor the test itself:

it.unibo.alchemist.test.PreventRegressions > testLoadAndSerialize() FAILED
org.apache.commons.lang3.SerializationException: java.io.NotSerializableException: it.unibo.alchemist.model.implementations.geometry.euclidean.twod.AwtEuclidean2DShapeFactory

**Comment by DanySK** [comment]
Well, needs to get debugged anyway. That test currently passes, so there is some change introduced here that breaks it.

**Comment by DanySK** [comment]
Can you reproduce it?

**Comment by mazzio97** [comment]
I'm trying to figure out what's going on. The test serialize and deserialize environments so it complains about the fact that AwtEuclidean2DShapeFactory, which is used inside EuclideanPhysics2DEnvironmentImpl, is not Serializable. Maybe putting Serializable on a bunch of classes can do the job but what's strange is that this kind of error must have been present since the creation of physics environments.

**Comment by mazzio97** [comment]
Ok maybe I've fixed it, now only one remains and it's even stranger...

it.unibo.alchemist.test.TestLoadGPSTrace > testLoadGPSTrace() FAILED
java.lang.IllegalArgumentException: Cannot create it.unibo.alchemist.model.implementations.actions.ReproduceGPSTrace with arguments [gpsTrace:String, true:Boolean, AlignToSimulationTime:String]

File: alchemist/alchemist-cognitive-agents/src/test/resources/feels-transmission-with-layer.yml [hunk 3]
```diff
@@ -5,7 +5,7 @@ variables:
     formula: "\"danger\""
 
 environment:
-  type: Continuous2DEnvironment
+  type: EuclideanPhysics2DEnvironmentImpl
```
[the remark this answers is not in the document]
Comment (mazzio97) [reply alchemist/alchemist-cognitive-agents/src/test/resources/feels-transmission-with-layer.yml]: Maybe Continuous2DEnvironment can become an interface extending PhysicsEnvironment like EuclideanPhysics2DEnvironment is now and EuclideanPhysics2DEnvironment can substitute EuclideanPhysics2DEnvironmentImpl. Obviously then more specific environments must extend EuclideanPhysics2DEnvironment rather than Continuous2DEnvironment.

[the remark this answers is not in the document]
Comment (mazzio97) [reply alchemist/alchemist-cognitive-agents/src/test/resources/feels-transmission-with-layer.yml]: Another option is to simply include EuclideanPhysics2DEnvironmentImpl code inside Continuous2DEnvironment and delete EuclideanPhysics2DEnvironmentImpl. Nothing else would be necessary.

[the remark this answers is not in the document]
Comment (DanySK) [reply alchemist/alchemist-cognitive-agents/src/test/resources/feels-transmission-with-layer.yml]: Both make sense. Think about it and pick the one you deem best.
Also, change the default to point to that environment, if needed.

**Comment by DanySK** [comment]
You have hardcoded paths.

**Comment by mazzio97** [comment]
We have talked about it, in order to load an ImageEnvironment there is no other way than specify the absolute path of the image. If I can do something different let me know.

**Comment by DanySK** [comment]
This is impossible. I do load from classpath in tests, and yes, we must do something different, it must pass on CI.
It is impossible that I consented to keeping hardcoded references of your file system in Alchemist. Not even under duress. ;)

**Comment by mazzio97** [comment]
Sorry, my fault, obviously tests can't pass the check with absolute paths. The problem we were talking about was for simulations running using the GUI, which since the classpath to provide is alchemist-main don't have a relative reference to a resource of some other module, so you said me to use absolute paths locally in order to deceive that problem.

**Comment by DanySK** [comment]
Within Idea you do what you please, but of course tests on the product must run in CI.
